### ISSUE TITLE
Bulk Processor

### ISSUE THREAD

**Issue by olivere** [olivere/elastic#190]
Create a bulk processor just like the one in the Java API.

See [BulkProcessor.java](https://github.com/elastic/elasticsearch/blob/2.x/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java) and its [documentation](https://www.elastic.co/guide/en/elasticsearch/client/java-api/current/java-docs-bulk-processor.html).

**Comment by sundarv85** [comment]
Thanks for adding the enhancement request. Do you have any timeline in mind, when this would be available. This would help us in our planning.

**Comment by olivere** [comment]
@sundarv85 I added a first draft of a `BulkProcessor` in the [bulk-processor](https://github.com/olivere/elastic/tree/bulk-processor) branch.

Notice that this is basically untested and most certainly not the final API. It's just a first draft. If you like to give it a try, or have time for code review: That'd be more than welcome. See the tests for usage examples.

**Comment by sundarv85** [comment]
@olivere Tested the bulk processor. It works good. Found an issue. 

When I have more than 1 worker (say 2), and I have set only periodic flushing interval (the size and action count is set to 0), then when the timer expires, only one worker is woken up. Because the `<-p.ticker.C: ` is processed only by one of the workers.

**Comment by olivere** [comment]
@sundarv85 I see. Thanks for testing. I will have a detailed look at it.

**Comment by olivere** [comment]
@sundarv85 I [just committed a new version](https://github.com/olivere/elastic/commit/8cdbcafb59afe958e1739c1094e9bd29cf22bc73) of BulkProcessor to the bulk-processor branch. It should fix the bug you found. The ticker for `flushInterval` is now handled on a separate goroutine (`flusher`). That goroutine will periodically ask the workers to flush their outstanding bulk requests. Can you please test it again?

**Comment by sundarv85** [comment]
@olivere Works perfectly. Thanks for the quick solution. When would you be merging them into the main branch.

**Comment by olivere** [comment]
@sundarv85 I'd like to do some clean up and have another look on the design. It's a quick hack as it is now, and I at least want to get the API right. E.g. there might be a need for a manual `Flush` as a kind of commit. If we leave it like this, the processor might throw away requests if the application skips early and forgets to `Close`.

If you need it badly, feel free to fork this branch and merge back to release-branch.v3 when bulk processor is finally merged.

**Comment by olivere** [comment]
@sundarv85 Can you please test the bulk processor branch one more time? It's not perfect, but if you give it a thumbs up, I think I'll merge with release-branch.v3.

**Comment by sundarv85** [comment]
@olivere Great. I tested, including the new Flush call and it works as expected. Thumbs Up for merging.

**Comment by sundarv85** [comment]
@olivere  I found an issue. When I have 5 worker threads, not all are correctly woken up on Flush or Flush Interval expiring.

Because, 
```
			for i := 0; i < p.numWorkers; i++ {
				p.flushCh <- struct{}{}
			}
```

Since we are writing into a common channel, sometimes the same worker thread processes the flushCh twice, there by skipping few other worker threads. In my example, when 5 workers were there, worker 1 was woken up once, worker 2 was woken up twice and worker 3 was woken up twice. Worker 0 and 4 were never woken up.

The best would be to give each worker its own flushCh.

**Comment by olivere** [comment]
@sundarv85 Thanks for testing. Will fix.

**Comment by olivere** [comment]
@sundarv85 Okay. Can we have another Go? :-) I've tested locally in one of my applications and AFAICS it's working fine. But I'd love to have more testers. [Here's the commit details](https://github.com/olivere/elastic/commit/1644ca9c547db3113d4def4782b7a1d8440b3b29).

I've refactored the workers into a separate class. Although tightly bound to a `BulkProcessor`, this makes some things more obvious (such as workers having their own service, channel etc.).

Some other notable issues: Flush is now synchronous, i.e. it waits for the workers to acknowledge the flush. This is necessary if you use a BulkProcessor in a standalone program. If you want to ensure that all requests are committed, you can do a manual Flush. If we don't wait for the ack, the program could still exit early.

Also: Callers are responsible to set the index and type on every bulk request. If you were using the [Bulk API](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html), you could set the index/type in the `action_and_meta_data` and only provide the (optional) source in the 2nd line of the bulk request. With BulkProcessor, you need to set the index and type in every single bulk request added to `BulkProcessor.Add`. The reason for this is that the workers have their own BulkService which isn't initialized with an index/type. While I could add setters for index/type on the BulkProcessor, I think this would be even more confusing for end-users of BulkProcessor.

**Comment by sundarv85** [comment]
@olivere Tested it again. In my tests, everything seems to work as expected.

On manual code review, I noticed that the same channel is used for getting the ack that Flush has been completed

```
		w.flushC <- struct{}{}
		<-w.flushC // wait for completion
```

Wouldn’t it create an issue when two different go routines call the Flush one after another. Then the 2nd routines, write into `w.flushC` will be taken up as ack for the 1st go routine and then 1st go routines end of flush will be taken up as ack for the 2nd go routine. So the 2nd go routine's Flush would not be executed. It would be better if we create an `flushAckC`. What do you think?

**Comment by olivere** [comment]
@sundarv85 The `flushC` should be private to the worker (see [here](https://github.com/olivere/elastic/blob/bulk-processor/bulk_processor.go#L181)). When you send a value to the channel, the worker should receive it and the caller (`Flush`) should block as the channel is unbuffered. Then, when the worker successfully flushed, it will send to the channel and that should unblock `Flush`. Or am I missing something?

**Comment by olivere** [comment]
@sundarv85 I think I see what you're after, so I added a 2nd channel for the ack. Also, the flush channel wasn't closed, so this should be updated as well ([commit](https://github.com/olivere/elastic/commit/35cced54a5e683751c175ac3380618b8cb1e3125)).

**Comment by sundarv85** [comment]
@olivere Perfect! I tested the new code base as well. Thumbs Up for merging :-) 

And thanks for acting so quickly to finish the feature within such a short time!

**Comment by sundarv85** [comment]
@olivere Am not sure if it is an issue, but when the commit throws an error, the `w.service.NumberOfActions()` count does not go down. 

For e.g. in my case, I had bulkActions set to 10. Unknowingly I deleted my index from sense. So my writes were failing with the error

`elastic: Error 400 (Bad Request): Can't specify parent if no parent field has been configured [type=illegal_argument_exception]`

And within this my requests were piling up above the bulkActions count. For every error, of course, we should not discard the data because some errors are recoverable. But should there be an upper limit on how many requests can be cached? This is more related to the `service` rather than `bulk processor`.

**Comment by olivere** [comment]
@sundarv85 Error handling is hard. In the other services, Elastic usually is passive in the sense that it leaves error handling and cleanup to the caller. It simply passes errors to the caller.

However, if `BulkProcessor` gets added to core, we are controlling the `BulkService` inside the workers and we're going asynchronous. From within the worker however, the callbacks should be triggered, e.g. the `Failure` callback and you should get notified of errors and handle accordingly. You could e.g. stop the processor and start over. Did the `Failure` callback got triggered in your situation?

The question is: What is the most sensitive way to behave for `BulkProcessor` in case something goes badly wrong? We could simply stop and retry, but we could still have uncommitted requests. `Add` could return an error when this happens, but it's to some extent the same as implementing a `Failure` handler. I was looking into the [Java API](https://github.com/elastic/elasticsearch/blob/2.x/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java), but it also doesn't do anything to protect you: It also simply invokes the [`Listener.afterBulk`](https://github.com/elastic/elasticsearch/blob/2.x/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java#L68) callback (as [defined here](https://github.com/elastic/elasticsearch/blob/2.x/core/src/main/java/org/elasticsearch/action/bulk/BulkRequestHandler.java)).

What would you consider a sensible response from `BulkProcessor` when something fatal happens?

**Comment by sundarv85** [comment]
@olivere yes, I agree that the leaving it for the users to handle errors is a more simpler way and fits together with the Go approach of `less magic`.

Your new changes for the Failure handler suits me well. 

In my case, the FailureHandler is called. Should there be an API to clear all the requests within the workers. Because after my FailureHandler is called, it is again constantly called for every interval flushing.

**Comment by olivere** [comment]
@sundarv85 Your last question is regarding cleanup in the error case, is it?

Well, I'm undecided what's the best way to achieve that. If it's a temporary failure, maybe BulkProcessor should tell the client about the problems via Failure handler but still keep trying. If it's a permanent failure, Elastic cannot do anything really.

We probably have to think about the failure story from the client side: What is the expected way of doing cleanup in the error case from the client-side? One way to do that is to call `Close` on the BulkProcessor. BulkProcessor would try, once again, to commit all outstanding requests, but would ultimately close the workers. Requests could be lost, but what can we do?

We could try to limit impact by blocking BulkProcessor/workers if errors occur. Blocking here refers to the ability to add new documents via `Add`; the call would return an error. BulkProcessor could then automatically retry and unblock once the dust has settled. But to be honest, that would result in far too much logic inside BulkProcessor.

I think the initial approach is to implement an error handler that would do the "right thing" when something unexpected happens. The definition of "right" is up to the application: Retry, panic, start over.

We should try to figure out if this approach to error handling is possible and "nice" from the client perspective. E.g. the failure handler does not have a reference to the bulk processor, so how can it stop. Maybe we should split the `BulkProcessorService` into a builder and a `BulkProcessor` that does the actual work when started with `Do`. That would also simplify the design.

Unfortunately, I will be in meetings throughout most of next week, so I won't be available to work on it for a few days.

**Comment by sundarv85** [comment]
@olivere yes, I agree that we should keep things simple (with less magic) and leave it out to the client to decide what to do.

And you are correct that the failure handler probably will need the bulkprocessor reference to cleanup. Fro now, I can use closure to stop the bulkprocessor in the failure handler and restart it.

**Comment by olivere** [comment]
@sundarv85 Did you actually try that kind of error handling/cleanup and it worked for you? Or is it just some kind of thought experiment ;-)

**Comment by sundarv85** [comment]
@olivere It was a thought experiment and I indeed tried it just now. I’m able to (of course) access the bulkprocessor reference, however, I’m not able to update that instance within that failure function safely, because during that time other goroutines are trying to Add entries into the bulk and it panics with error `panic: send on closed channel`.

Also, within failure handler, I would not be able to re-register the same failure handler again. 

The only way would be to give a `Reset()` function in bulkprocessor and the FailureHandler can use that function to reset everything.

Instead of that, probably your approach of `Maybe we should split the BulkProcessorService into a builder and a BulkProcessor that does the actual work when started with Do. That would also simplify the design.` would be better.

**Comment by olivere** [comment]
@sundarv85 Thanks for testing. Yeah, the design is just not right. We should just let it sink in and will find a better approach. No reason to rush. :-)

**Comment by olivere** [comment]
@sundarv85 Searching for best practices on using the BulkProcessor in Java yields some interesting issues, e.g. [BulkProcessor - things to avoid](http://elasticsearch.narkive.com/xSGFIktv/bulkprocessor-nonodeavailableexception-things-to-avoid).

**Comment by olivere** [comment]
[BulkProcessor.java](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/action/bulk/BulkProcessor.java) in Elasticsearch master has changed recently, and there's discussion on changing the listener interface to be item-based as well. It now comes with a backoff-policy.

**Comment by sundarv85** [comment]
@olivere yes, lets just let it sink. And your findings in java BulkProcessor are interesting. We should probably monitor those changes as well. Meanwhile, I was wondering whether on every Failure, if the service doesn’t retry but just clears the buffer, if a Failure Fn is registered - would that be correct. So anyone who registers a failure function has to either requeue the data or can drop it.

It is more simpler approach. What do you think?

**Comment by olivere** [comment]
@sundarv85 Just to document the behavior: What `BulkService` does is to [reset](https://github.com/olivere/elastic/blob/bulk-processor/bulk.go#L167) if the bulk request has been successfully sent to Elasticsearch (i.e. no errors returned via HTTP response or deserializing JSON). Resetting means: Clear the list of bulk requests and the estimated byte size (see [here](https://github.com/olivere/elastic/blob/bulk-processor/bulk.go#L40)). The failure handler gets the requests and the response when an error occured. In that case, `reset` is not called. So the bulk requests in the service don't get cleared.

You are asking `BulkProcessor` to reset anyway and make it up to the caller to re-add requests to it?

**Comment by sundarv85** [comment]
@olivere yes, that is correct. I was thinking that the `BulkProcessor` can do a reset on failure case as well, to make it up to the caller to re-add requests if necessary.

**Comment by sundarv85** [comment]
@olivere For the new v3 release, can you merge the bulk processor branch as well.

**Comment by olivere** [comment]
@sundarv85 I'd really like to have a good story for "how to handle errors for the client-side" first. I tested something on the weekend, but it's not as nice as it could be yet. E.g. you could throttle or even stop the client to hand over new requests via `Add` by introducing a new channel. It will trigged throttling in `Failure` handler, and will lift the throttling again in the `After` handler. Combined with `FlushInterval` (which takes care of the retries), this is a viable solution. But I think we could do better.

What I could do is to extract the changes to the Bulk API and commit it into the release branches, i.e. the methods to get the estimated size of a bulk commit. With that change, it should be possible to take the `BulkProcessor` from the `bulk-processor` branch and use it internally until `BulkProcessor` is stable and can be moved into the core of Elastic.

**Comment by olivere** [comment]
@sundarv85 Hey. Can you give BulkProcessor another shot, test- and "code review"-wise? 

Here's a brief summary of the most important changes:

1. The workers now come with retries via exponential backoff, i.e. they try several times before giving up.
2. Failure and After handler merged into one, leaving only Before and After. After now has an error parameter that indicates if the last batch returned an error or not.
3. I've documented the current state of [`BulkProcessor` in the Wiki](https://github.com/olivere/elastic/wiki/BulkProcessor). That includes an example program (see [the gist at the end](https://gist.github.com/olivere/a1dd52fc28cdfbbd6d4f)).

Thanks in advance!

**Comment by olivere** [comment]
@sundarv85 I just pushed `BulkProcessor` in 3.0.19 and marked it as experimental.

**Comment by sundarv85** [comment]
@olivere Sorry! I got held up this week. We are working towards a release and I’m held up this weekend as well. I will test it again and update you early next week. Thanks.

**Comment by olivere** [comment]
@sundarv85 No problem. Thanks for you commitment.

**Comment by olivere** [comment]
@sundarv85 Hi there. Did you find some time to look into it yet?

**Comment by jgoodall** [comment]
I have done some testing and have found that it is necessary to adjust some of the `threadpool.bulk` settings in elasticsearch. If this is not done, we lose data silently (i.e. no errors are thrown by the `BulkProcessor`).

`BulkProcessor` configuration:
```go
const bulkActions = 1000               // commit if # requests >= 1,000
const bulkSize = 5 << 20               // commit if size of requests >= 5 MB

bulkService, err := client.BulkProcessor().
	Workers(2).
	BulkActions(bulkActions).
	BulkSize(bulkSize).
	Do()

```

`elasticsearch.yml`:

```yml
threadpool.bulk.type: fixed
threadpool.bulk.queue_size: 1500
```

It might be helpful to put this in the wiki...

**Comment by olivere** [comment]
Okay. Silently losing data is not acceptable. Can you provide more insights into how your tests work? E.g. did the failure callbacks get triggered, and did they return failed requests?

**Comment by jgoodall** [comment]
No, after callback was called, but err was never non nil. Only way to tell is checking failed in the stats.



________________________________

**Comment by olivere** [comment]
Ah. So Failed in stats was > 0?

**Comment by jgoodall** [comment]
Yes. Exactly. But not A consistent number across runs.


________________________________

**Comment by sundarv85** [comment]
@olivere Sorry for the long delay. I just tested and everything seems to be working fine.

@jgoodall Did you call the `BulkProcessor Close()` before your program exits. If you don’t call the Close() or Flush(), then the data may be dropped silently.

**Comment by jgoodall** [comment]
@sundarv85 Yes, I flush then close. The dropped documents actually happens much earlier (in my tests I have 23 million documents, and before the first million have been indexed some are dropped, unless upping the default `threadpool.bulk.queue_size`):

```go
	// Watch for SIGINT and SIGTERM and clean up before exiting.
	go func() {
		c := make(chan os.Signal)
		signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
		<-c
		fmt.Println("Exiting...")
		bulk.Flush()
		bulk.Close()
		os.Exit(0)
	}()
```

**Comment by olivere** [comment]
@jgoodall I see. Having failed documents seems to be non-exceptional case to me. Note that "failed" doesn't mean that Elastic didn't try to bulk insert/create/update/delete the request. It just means that ES marked the given request as "failed" in the response. That might happen for various reasons. One of them is, probably, an overloaded ES cluster. Hence upping the limits on the bulk threadpool fixed it for you.

It is the responsibility of the application to handle those requests, e.g. by re-adding them to the bulk processor, triggering an event, putting it into a dead letter queue etc. This could also happen if you're using the bulk feature outside of bulk processor.

Note that in your example, `Close` should automatically `Flush`, so you should only need `Close`.

And you're right, of course, that all these things should be documented in the Wiki. Thanks for testing btw :-)

**Comment by sundarv85** [comment]
@jgoodall Could you get the stats of the thread_pool of bulk. You will see if there were failed requests:

 `GET /_nodes/stats/thread_pool`

**Comment by olivere** [comment]
@sundarv85 Hey, thanks for testing once again. I've been using it in my apps for a while, and it worked fine. But I didn't do anything near the 23 mio. that @jgoodall tested with. I think we could consider this feature done then. Thanks again for your support.

**Comment by olivere** [comment]
@sundarv85 @jgoodall I added a [FAQ section](https://github.com/olivere/elastic/wiki/BulkProcessor#faq) to the bulk processor wiki page.

**Comment by sundarv85** [comment]
@olivere Yep! I agree. This can be considered closed.

**Comment by jgoodall** [comment]
@olivere If I remove the threadpool settings and rerun, here is what Elasticsearch reports (`GET /_nodes/stats/thread_pool`):

```json
        "bulk" : {
          "threads" : 4,
          "queue" : 0,
          "active" : 0,
          "rejected" : 2549,
          "largest" : 4,
          "completed" : 259239
        },
```

Here is what the `BulkProcessor` stats report:

```
"Committed": 23309,
"Created": 16738391,
"Failed": 79407,
"Flushed": 109,
"Indexed": 6488687,
"Succeeded": 23147671,
```

A couple of followup questions:

* Just to confirm - if a commit in `BulkProcessor` fails, it will retry using `ExponentialBackoff` before throwing an error correct?

* Your comment says the application needs to handle the failures, but I am having trouble figuring out how to. `err` is `nil` in the `After` callback - so how can I catch and retry the errors?

**Comment by olivere** [comment]
@jgoodall Okay, good points for further documentation.

Exponential backoff is only tried with "hard" errors, i.e. when the [Bulk API returns an error](https://github.com/olivere/elastic/blob/release-branch.v3/bulk_processor.go#L470). An error here is defined by ES returning a HTTP status outside the [200,300) range.

However, this is not the same as a document that has been rejected by Elasticsearch. An example is when you try to delete a document via [`BulkDeleteRequest`](https://github.com/olivere/elastic/blob/release-branch.v3/bulk_delete_request.go) and that document doesn't exist. When you do that, ES will return HTTP status 200 and return a large response structure that describes the outcome of every single bulk request you sent, including the given `BulkDeleteRequest` that has failed.

Notice that this can happen all the time when using the Bulk API. You should always inspect the response of the Bulk API and see if all documents have been processed as expected:

```go
res, err := bulkService.Do()
if err != nil {
   ...
}
failedRequests := res.Failed()  // returns a []*elastic.BulkResponseItem
for _, req := range failedRequests {
  // req.Id contains the Id of the failed document
}
```

There are various [helper functions](https://github.com/olivere/elastic/blob/release-branch.v3/bulk.go#L231) to inspect the response and e.g. filter out those that failed. There's also some [tests](https://github.com/olivere/elastic/blob/release-branch.v3/bulk_test.go).

Now, how do you inspect the response in bulk processor? You can do that by specifying a callback via the [`After`](https://github.com/olivere/elastic/blob/release-branch.v3/bulk_processor.go#L64) function. Here's an example:

```go
afterFn := func(executionId int64, requests []elastic.BulkableRequest, response *elastic.BulkResponse, err error) {
  if err != nil {
    // Something went badly wrong, ES reported HTTP status outside [200,300), even after retrying
    return // response is probably nil
  }
  failed := response.Failed()
  if len(failed) > 0 {
    // Handle failed requests appropriately
  }
}

p, err := elastic.NewBulkProcessor().
  After(afterFn).
  Do()
...
```

Hope that helps?!

**Comment by olivere** [comment]
Closing as there is no activity recently. Reply or create a new issue if required.
